include/seqan3/range/decorator/gap_decorator_anchor_set.hpp

```diff
@@ -0,0 +1,485 @@
+    constexpr reference operator[](size_type const i) const
+    {
+        assert(i < size());
+        // case 1: no gaps
+        if (!anchors.size()) return value_type((*sequence)[i]);
+        // case 2: gaps
+        anchor_set_iterator it = anchors.lower_bound(gap_t{i, _});
+        if (anchors.size() && it == anchors.end())
+            it = std::prev(it);
+
+        size_type acc = 0, gap_len = 0;
+        if ((*it).first <= i || (it != anchors.begin() && (*(std::prev(it))).first <= i))
+        {
+            if ((*it).first > i)    --it;
+            acc = (*it).second;
+            gap_len = (*it).second;
+            if (*it != *(anchors.begin()))
+                gap_len -= (*(std::prev(it, 1))).second;
+            if (i >= (*it).first && i < (*it).first + gap_len)
+                return gap::GAP;
+        }
+        return value_type((*sequence)[i - acc]);
+    }
+
+    /*!\name Comparison operators
+     * \brief Compare gap decorators by underlying sequence and gaps.
+     * \{
+     */
+    template <typename inner_type2>
+    //!\cond
+        requires std::is_same_v<std::remove_const_t<inner_type>, std::remove_const_t<inner_type2>>
+    //!\endcond
+    // TODO: input gap_decorator could be of any subtype, e.g. anchor_set, anchor_list, etc.
+    constexpr bool operator==(gap_decorator_anchor_set<inner_type2> const & rhs) const noexcept
+    {
+        if (this->anchors.size() != rhs.anchors.size() || this->sequence->size() != rhs.sequence->size())
+            return false;
+
+        return std::ranges::equal(*this, rhs);
+    }
+
+    template <typename inner_type2>
+    //!\cond
+        requires std::is_same_v<std::remove_const_t<inner_type>, std::remove_const_t<inner_type2>>
+    //!\endcond
+    constexpr bool operator!=(gap_decorator_anchor_set<inner_type2> const & rhs) const noexcept
+    {
+        return !(this->operator==(rhs));
+    }
+
+private:
+    //!\privatesection
+    /*!\brief Helper function to compute the length of the gap indicated by the
+    * input iterator.
+    * \details The length of a gap is difference of the accumulator pointed at
+    * and the one of its predecessor (if existing).
+    */
+    constexpr size_type get_gap_length(anchor_set_iterator it) const noexcept
+    {
+        if (it == anchors.begin()) return (*it).second;
+        return (*it).second - (*std::prev(it)).second;
+    }
+
+    /*!\brief Update all anchor gaps after the indicated position by adding an offset.
+    * \details For not invalidating the iterator over the ordered set, the
+    * update is done in reverse manner excluding the indicated gap.
+    */
+    void rupdate(size_type const pos, size_type const size)
+    {
+        size_type new_key, new_val;
+        for (auto it = std::prev(anchors.end(), 1); (*it).first > pos;)
+        {
+            new_key = (*it).first + size;
+            new_val = (*it).second + size;  //idx2len[*it] + size;
+            anchors.insert(gap_t{new_key, new_val});
```

Comment:
again use emplace_hint

Reply:
Yes, here it should work fine since `size` is guaranteed to be larger than 0, so no key collisions!